File: iosApp/iosApp/Features/Desserts/Review/ReviewFormViewModel.swift
//
//  ReviewFormViewModel.swift
//  iosApp
//
//  Created by Michael Stromer on 1/30/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared

class ReviewFormViewModel: ViewModel, ObservableObject {
    @Published var review: Review?
    
    let reviewRepository: ReviewRepository
    
    init(reviewRepository: ReviewRepository) {
        self.reviewRepository = reviewRepository
    }
    
    func createReview(dessertId: String, newReview: Review) {
        reviewRepository.doNewReview(dessertId: dessertId, reviewInput: ReviewInput(rating: Int32(newReview.rating), text: newReview.text)) { [weak self] (data, error) in
            guard let self = self,
                  let newDessert = data else { return }
            self.review = newDessert
        }
    }
    
    func updateReview(review: Review) {
        reviewRepository.updateReview(reviewId: review.id, reviewInput: ReviewInput(rating: Int32(review.rating), text: review.text)) { [weak self] (data, error) in
            guard let self = self,
                  let updatedDessert = data else { return }
            self.review = updatedDessert
        }
    }
    
    func deleteReview(reviewId: String) {
        reviewRepository.deleteReview(reviewId: reviewId) { [weak self] (deleted, error) in
            guard let self = self else { return }
            self.review = nil
        }
    }
}

File: androidApp/src/main/java/com/example/justdesserts/androidApp/ui/desserts/review/ReviewFormViewModel.kt
package com.example.justdesserts.androidApp.ui.desserts.review

import androidx.lifecycle.ViewModel
import com.apollographql.apollo.api.ApolloExperimental
import com.example.justdesserts.shared.cache.Dessert
import com.example.justdesserts.shared.cache.Review
import com.example.justdesserts.shared.repository.ReviewRepository
import com.example.justdesserts.type.DessertInput
import com.example.justdesserts.type.ReviewInput

@ApolloExperimental
class ReviewFormViewModel constructor(private val repository: ReviewRepository): ViewModel() {
    suspend fun getReview(reviewId: String): Review? {
        return repository.getReview(reviewId)
    }
    suspend fun createReview(dessertId: String, review: Review): Review? {
        return repository.newReview(dessertId, ReviewInput(text = review.text, rating = review.rating.toInt()))
    }
    suspend fun updateReview(review: Review): Review? {
        return repository.updateReview(review.id, ReviewInput(text = review.text, rating = review.rating.toInt()))
    }
    suspend fun deleteReview(reviewId: String): Boolean? {
        return repository.deleteReview(reviewId)
    }
}